src/test/java/org/breedinginsight/brapps/importer/ExperimentFileImportTest.java

```diff
@@ -0,0 +1,885 @@
+        assertEquals(expected.get(Columns.ENV_LOCATION), Utilities.removeProgramKey(location.getName(), program.getKey()));
+        assertEquals(expected.get(Columns.ENV_YEAR), season.getSeasonName());
+        assertEquals(expected.get(Columns.EXP_UNIT_ID), Utilities.removeProgramKeyAndUnknownAdditionalData(ou.getObservationUnitName(), program.getKey()));
+
+        BrAPIObservationUnitLevelRelationship rep = null;
+        BrAPIObservationUnitLevelRelationship block = null;
+        for (BrAPIObservationUnitLevelRelationship rel : ou.getObservationUnitPosition().getObservationLevelRelationships()) {
+            if ("rep".equals(rel.getLevelName()) && rep == null) {
+                rep = rel;
+            } else if ("block".equals(rel.getLevelName()) && block == null) {
+                block = rel;
+            }
+        }
+        assertNotNull(rep);
+        assertNotNull(block);
+        assertEquals(expected.get(Columns.REP_NUM), rep.getLevelCode());
+        assertEquals(expected.get(Columns.BLOCK_NUM), block.getLevelCode());
+        if(expected.containsKey(Columns.ROW)) {
+            assertEquals(expected.get(Columns.ROW), ou.getObservationUnitPosition().getPositionCoordinateX());
+            assertEquals(BrAPIPositionCoordinateTypeEnum.GRID_ROW, ou.getObservationUnitPosition().getPositionCoordinateXType());
+        }
+        if(expected.containsKey(Columns.COLUMN)) {
+            assertEquals(expected.get(Columns.COLUMN), ou.getObservationUnitPosition().getPositionCoordinateY());
+            assertEquals(BrAPIPositionCoordinateTypeEnum.GRID_COL, ou.getObservationUnitPosition().getPositionCoordinateYType());
+        }
+        if(expected.containsKey(Columns.TREATMENT_FACTORS) && StringUtils.isNotBlank((String)expected.get(Columns.TREATMENT_FACTORS))) {
+            assertEquals(expected.get(Columns.TREATMENT_FACTORS), ou.getTreatments().get(0).getFactor());
+        }
+
+        if(traits != null) {
+            List<String> expectedVariableObservation = new ArrayList<>();
+            List<String> actualVariableObservation = new ArrayList<>();
+            observations.forEach(observation -> actualVariableObservation.add(String.format("%s:%s", Utilities.removeProgramKey(observation.getObservationVariableName(), program.getKey()), observation.getValue())));
+            for(Trait trait : traits) {
+                expectedVariableObservation.add(String.format("%s:%s", trait.getObservationVariableName(), expected.get(trait.getObservationVariableName())));
+            }
+
+            assertThat("Missing Variable:Observation combo", actualVariableObservation, containsInAnyOrder(expectedVariableObservation.toArray()));
+        }
+
+        return ret;
+    }
+
+    private Map<String, Object> assertValidPreviewRow(Map<String, Object> expected, JsonObject actual, Program program, List<Trait> traits) {
```

Comment:
is this private method ever called

Reply:
Not at the moment.  I was thinking that there should probably be some tests added for preview of import (current tests are only focused on the full save)

Reply:
Makes sense